### PR TITLE
[#5874] [#5971] Ensure SniHandlerTest.testServerNameParsing not fails…

### PR DESCRIPTION
**Pull request by normanmaurer**
… with SslProvider.JDK

Motivation:

The SniHandlerTest.testServerNameParsing did fail when SslProvider.JDK was used as it the JDK SSLEngineImpl does not send an alert.

Modifications:

Ensure tests pass with JDK and OPENSSL ssl implementations.

Result:

SniHandlerTest will run with all SslProvider and not fail when SslProvider.JDK is used.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler PTAL

**Comment by Scottmitch** [comment]
Is it not a problem that the OpenSslEngine and JdkSslEngine have different behaviors? I looked at this briefly before but I didn't have time to determine if the way were were handling errors was different for the two engines, if the error should have originated from SSLEngineImpl and didn't, or if the spec allows flexibility here for an alert not to be generated.

File: handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java [hunk 1]
```diff
@@ -75,292 +80,348 @@ private static ApplicationProtocolConfig newApnConfig() {
...
         File keyFile = new File(SniHandlerTest.class.getResource("test_encrypted.pem").getFile());
         File crtFile = new File(SniHandlerTest.class.getResource("test.crt").getFile());
 
-        return SslContextBuilder.forServer(crtFile, keyFile, "12345")
-                .sslProvider(provider)
-                .applicationProtocolConfig(newApnConfig()).build();
+        SslContextBuilder sslCtxBuilder = SslContextBuilder.forServer(crtFile, keyFile, "12345")
+                .sslProvider(provider);
+        if (apn) {
+            sslCtxBuilder.applicationProtocolConfig(newApnConfig());
```
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java]: should we do the same `assumeTrue` checks as we do in `makeSslContext` here?

File: handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java [hunk 2]
```diff
@@ -75,292 +80,348 @@ private static ApplicationProtocolConfig newApnConfig() {
...
+        SslContextBuilder sslCtxBuilder = SslContextBuilder.forClient().trustManager(crtFile).sslProvider(provider);
+        if (apn) {
+            sslCtxBuilder.applicationProtocolConfig(newApnConfig());
+        }
+        return sslCtxBuilder.build();
+    }
+
+    @Parameterized.Parameters(name = "{index}: sslProvider={0}")
+    public static Iterable<? extends Object> data() {
+        List<SslProvider> params = new ArrayList<SslProvider>();
```
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java]: initial size of 3?

File: handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java [hunk 3]
```diff
@@ -75,292 +80,348 @@ private static ApplicationProtocolConfig newApnConfig() {
...
+                        sc.close().syncUninterruptibly();
+                    }
+                    if (sslContext != null) {
+                        ReferenceCountUtil.release(sslContext);
+                    }
+                    group.shutdownGracefully();
 
-            cert.delete();
+                    cert.delete();
+                }
```
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java]: add a `case JDK: return;` and `case default: throw new Error();`?

File: handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java (hunk 2 above)
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java]: sure I can do... but honestly I don't think it matters... Its a unit test.

File: handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java (hunk 3 above)
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java]: yep why not.

**Comment by normanmaurer** [comment]
@Scottmitch I think what we do in openssl is correct. So I think its either an SSLEngineImpl bug or its optional. So I would just not put an assert there.

File: handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java (hunk 1 above)
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java]: This is part of the `makSslContext` method and so it already has the guards.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (86bbf24) and 4.0 (d80bd8fb4aa04490171e398b1cd17fbf0237815b)

**Pull request closed without merging**